### ISSUE TITLE
Crashing after launch

### ISSUE THREAD

**Issue by acvilonre** [jpxue/Overwatch-Aim-Assist#2]
Right after the program start it crashes 
http://imgur.com/a/me3T8

**Comment by jpxue** [comment]
That is very odd indeed, the method that I'm using for capturing a screenshot is a pretty standard one, but from that info I can't tell why GetObject is causing an exception.
Run the application in debug mode rather than release and tell me what the error is.

I suspect that you're getting "error reading registry value" because it's running in release mode. and hence the local values are not being reported back to the debugger due to optimizations; hence the 'value is optimized away and not available' messages.

TL;DR: Run it in debug mode and report back to me.

Thanks!

**Comment by ParasiteEve** [comment]
http://i.imgur.com/6NQOhVL.png

This is the crash I am getting, it's crashing for me as well.

**Comment by jpxue** [comment]
Thank you for the info ParasiteEve.

Looks like the problem arises when trying to dereference a null pointer. Try modifying line 66 to the following:

```
if(pixels!=NULL)
        delete[] pixels; 

```

Or you can download the source again and report back to me please.

**Comment by ParasiteEve** [comment]
http://i.imgur.com/T8CSq6h.png

Added the code; this is what occurs, same error. Thanks for the help.

**Comment by jpxue** [comment]
Thanks again for your reply.
Hmm, I'm pretty puzzled to be honest as I cannot seem to find a way to replicate the error for investigation purposes. I am assuming that it occurs on starting the program, specifically when line 101 (in Capture.cpp) is called for the first time.

Does the program crash on a pre-compiled version?:
https://www.mediafire.com/?a3omopoh1k20cm1

**Comment by ParasiteEve** [comment]
Might be due to my version of VS? I am using 2015. I googled and this may be the case. What version do you use? I will test the pre compiled when I get home.

**Comment by jpxue** [comment]
I am using VS2015 Pro edition but as you pointed out, yes, some people have solved this by changing VS versions which I find very odd, hence why I sent you a compiled version (to test this possibility out).
Thanks

**Comment by acvilonre** [comment]
It crashes in release but when I debug it starts work but it lags very much

**Comment by kkm** [comment]
I am using VS2013 and .Net 4.5.
In Release or Debug both work fine for me.

**Comment by jpxue** [comment]
wasdasf does my pre-compiled version work for you? https://www.mediafire.com/?a3omopoh1k20cm1

I genuinely have no idea why it would run on debug (which is rightfully laggy) and not on release.
I need more info for instance the exception details when running in release; I know that it is crashing on GetObject but I don't really know the specifics.

**Comment by acvilonre** [comment]
The precompiled version runs but getting error about blackscreen and is still laggy and yes i have it on borderless

**Comment by jpxue** [comment]
I can confirm that the black screenshot thingy is an issue for **some** people yes (because I received 3 private messages about it).
I currently don't know why some people are getting a black screenshot, even in borderless windowed mode but I will look into it tomorrow if I have the time and motivation.

As for the lag, I cannot understand how it  is lagging if it does nothing but report that the screenshot is black and waits for half a second before taking another screenshot and checking it again:
`while (!recorder.screenshotGDI(screeny)) Sleep(500);`

**Comment by ParasiteEve** [comment]
Hi JP, thanks for the support. Precompiled works but I do get the black screen error and also it completely aims in the wrong direction when it sees health bar. I'd edit the code a bit but I can't recompile due to the error. Will try 2013 I suppose.

**Comment by ParasiteEve** [comment]
http://i.imgur.com/Nfzl7Nz.png
http://i.imgur.com/yMYk4CW.png
http://i.imgur.com/YVk7UZL.png
http://i.imgur.com/Y81gOFk.png
http://i.imgur.com/VgrQfpg.png

Installed 2013 and same errors. I looked at the call stack and it shows 5 possible locations that's causing the error, I have put the screenshot up along with highlighting to indicate it.

**Comment by ParasiteEve** [comment]
When I rebuild the solution there's this warning:
1>  Screenshot.cpp
1> screenshot.cpp(109): warning C4244: 'initializing' : conversion from 'float' to 'int', possible loss of data

**Comment by jpxue** [comment]
I don't know how you're getting a black screen error yet it still tries to aim, it should be stuck taking endless screenshots until one is obtained successfully.

We do not really need precision in your implementation of line 109 but you can typecast if it bothers you by putting (int) before percentagewidth. 
PS: I noticed that you're using a previous commit/version, try and download the latest changes where I removed some unnecessary things/constructors (which MIGHT have been leading a heap corruption in some users).

**Comment by jpxue** [comment]
I re-implemented the screenshot function after googling why GetDIBits returns all black pixels.
Can someone test the changes out before resorting to another approach?

**Comment by acvilonre** [comment]
I think Ive figured out why the program lags while have both overwatch and the cheat opened Overwatch is use 32% of my cpu and 1,182MB of my memory with the cheat it self also using roughly 14% of my cpu. Which for me is extremely odd and of course you gotta take in account my computer is extremely bad to compared to your I bet. So I don't know if there is a way to optimize the cheat for crappier computers, but if so would greatly appreciate.

**Comment by jpxue** [comment]
If by lag, you mean jittery mouse movements then I'd recommend limiting the FPS to display mode; it will do wonders!

Hmm, my CPU is 6 years old and Overwatch uses 15-20%/1.1GB of my CPU & RAM respectively.
The program uses 7%/15mb in my case so yeah, I can imagine it being laggy on your machine if it's consuming 14% of your CPU.

I could potentially run Overwatch in a VM so as to simulate the processing power of really crappy computers and optimize the code (_maybe I need more sleep functions so as to tax the CPU a little bit less or use threads to increase performance_); idk unless I try this out but I currently do not have time to write code.

At least, I'm assuming that the black screenshot issue was solved seeing that I stopped receiving PMs about it too, but I'd rather have confirmation.

**Comment by acvilonre** [comment]
Yeah you fixed the black screenshot issue

**Comment by kkm** [comment]
@ParasiteEve All files of *.h should into folder Header Files.

**Comment by ParasiteEve** [comment]
The new commit fixes all the errors I had, thanks. Works fine now. There are obvious issues such as multiple enemy bars and whole screen capture but the core works

**Comment by acvilonre** [comment]
You can change the % of your screen it captures
